aten/src/ATen/core/op_registration/README.md

```diff
@@ -0,0 +1,239 @@
+# Howto: Writing PyTorch & Caffe2 Operators
+
+So you want to write a new operator or a new kernel for an existing operator. How do you do that and what API should you use? So glad you asked.
+
+## native_functions.yaml vs custom operators
+
+All operators that are part of the public API of PyTorch are defined in `native_functions.yaml`. Just add an entry there and write the corresponding C++ kernel function. It’s very easy and there is a good introduction at https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md.
+
+### So when should you **not** use `native_functions.yaml`?
+
+There’s four main use cases
+
+* You’re writing a new operator that isn’t supposed to be part of the public PyTorch API.
+* You’re writing a new operator but don’t want to change the core pytorch code base, say you’re developing a shared library with operators.
+* You’re writing a C++ extension for PyTorch or you’re using inline c++ in your .py model files.
+* You’re writing a backend library like XLA or MSNPU that adds new kernels to all operators defined in `native_functions.yaml`.
+
+For these use cases, the custom operator API is the better solution.
+
+### What is the price for using the custom operator API instead of `native_functions.yaml`?
+
+If you’re just using the custom operator API to add new kernels for existing operators (e.g. the XLA/MSNPU example above), then you’re fine and don’t pay any price. If, however, you define a new operator purely using the custom op API, i.e. your operator never shows up in `native_functions.yaml`, then you need to be aware of a few caveats.
+
+* It will not get a C++ API generated. There will not be `Tensor::your_op()` methods or `at::your_op()` functions to call your operator.
+* The API for calling the operator from Python looks a little bit different. It needs to be called through `torch.ops.your_op()` instead of `torch.nn.functional`.
+* Setting up autograd for custom operators is harder. You don’t get it automatically but need to use `torch::autograd::Function` to implement autograd. Note also that `torch::autograd::Function` does not work together with dispatch, so if you have different kernels for different backends (say CPU and CUDA), you need to manually write if/else statements for that.
```

Comment:
maybe add a small example of autograd function? (or do we have some doc for it?)

Reply:
We have tests for custom autograd functions (e.g. https://github.com/pytorch/pytorch/blob/d762ad09df7f5808196b0e2e417b6592e0d30a30/test/cpp/api/autograd.cpp#L126-L152) that might be useful (we don't have public docs for it yet :/ )

Reply:
added a link to that